common/kernel/sdc.cc

```diff
@@ -0,0 +1,453 @@
+/*
+ *  nextpnr -- Next Generation Place and Route
+ *
+ *  Copyright (C) 2020  gatecat <gatecat@ds0.me>
+ *  Copyright (C) 2024  rowanG077 <goemansrowan@gmail.com>
+ *
+ *
+ *  Permission to use, copy, modify, and/or distribute this software for any
+ *  purpose with or without fee is hereby granted, provided that the above
+ *  copyright notice and this permission notice appear in all copies.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ *  WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ *  MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ *  ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ *  WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ *  ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ *  OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *
+ */
+
+#include "log.h"
+#include "nextpnr.h"
+
+#include <algorithm>
+#include <iterator>
+
+NEXTPNR_NAMESPACE_BEGIN
+
+struct TclEntity
+{
+    enum EntityType
+    {
+        ENTITY_CELL,
+        ENTITY_PORT,
+        ENTITY_NET,
+        ENTITY_PIN,
+    } type;
+    IdString name;
+    IdString pin; // for cell pins only
+
+    TclEntity(EntityType type, IdString name) : type(type), name(name) {}
+    TclEntity(EntityType type, IdString name, IdString pin) : type(type), name(name), pin(pin) {}
+
+    const std::string &to_string(Context *ctx) { return name.str(ctx); }
+
+    CellInfo *get_cell(Context *ctx) const
+    {
+        if (type != ENTITY_CELL)
+            return nullptr;
+        return ctx->cells.at(name).get();
+    }
+
+    PortInfo *get_port(Context *ctx) const
+    {
+        if (type != ENTITY_PORT)
+            return nullptr;
+        return &ctx->ports.at(name);
+    }
+
+    NetInfo *get_net(Context *ctx) const
+    {
+        if (type == ENTITY_PIN) {
+            CellInfo *cell = nullptr;
+            if (ctx->cells.count(name)) {
+                cell = ctx->cells.at(name).get();
+            } else {
+                const std::string &n = name.str(ctx);
+                auto pos = n.rfind('.');
+                if (pos == std::string::npos)
+                    return nullptr;
+                // remove one hierarchy layer due to Radiant weirdness around PLLs etc
```

Comment:
this is a nexus-specific workaround that I don't think needs to be kept in the generic code